File: packages/server/src/services/environment-manager.ts
import type { CLIType } from '@opencode-autopilot/shared';

interface EnvironmentConfig {
  inherit: boolean;
  variables: Record<string, string>;
  secrets: string[];
  passthrough: string[];
}

interface SessionEnvironment {
  sessionId: string;
  cliType: CLIType;
  config: EnvironmentConfig;
  resolved: Record<string, string>;
}

const DEFAULT_PASSTHROUGH_VARS = [
  'PATH',
  'HOME',
  'USER',
  'SHELL',
  'TERM',
  'LANG',
  'LC_ALL',
  'TMPDIR',
  'TMP',
  'TEMP',
  'NODE_ENV',
  'BUN_ENV',
];

const CLI_SPECIFIC_VARS: Record<CLIType, string[]> = {
  opencode: ['OPENCODE_*', 'ANTHROPIC_API_KEY', 'OPENAI_API_KEY'],
  claude: ['ANTHROPIC_API_KEY', 'CLAUDE_*'],
  aider: ['AIDER_*', 'OPENAI_API_KEY', 'ANTHROPIC_API_KEY', 'GEMINI_API_KEY'],
  cursor: ['CURSOR_*', 'OPENAI_API_KEY'],
  continue: ['CONTINUE_*', 'OPENAI_API_KEY', 'ANTHROPIC_API_KEY'],
  cody: ['SRC_*', 'CODY_*'],
  copilot: ['GITHUB_*', 'GH_*'],
  custom: [],
  adrenaline: ['ADRENALINE_*', 'OPENAI_API_KEY'],
  'amazon-q': ['AWS_*', 'Q_*'],
  'amazon-q-developer': ['AWS_*', 'Q_*'],
  'amp-code': ['AMP_*'],
  auggie: ['AUGGIE_*'],
  'azure-openai': ['AZURE_*', 'OPENAI_*'],
  bito: ['BITO_*'],
  byterover: ['BYTEROVER_*'],
  'claude-code': ['ANTHROPIC_API_KEY', 'CLAUDE_*'],
  'code-codex': ['CODEX_*'],
  codebuff: ['CODEBUFF_*'],
  codemachine: ['CODEMACHINE_*'],
  codex: ['CODEX_*', 'OPENAI_API_KEY'],
  crush: ['CRUSH_*'],
  dolt: ['DOLT_*'],
  factory: ['FACTORY_*'],
  gemini: ['GEMINI_API_KEY', 'GOOGLE_API_KEY'],
  goose: ['GOOSE_*'],
  grok: ['GROK_API_KEY', 'XAI_API_KEY'],
  jules: ['JULES_*'],
  'kilo-code': ['KILO_*'],
  kimi: ['KIMI_API_KEY', 'MOONSHOT_API_KEY'],
  llm: ['LLM_*', 'OPENAI_API_KEY'],
  litellm: ['LITELLM_*', 'OPENAI_API_KEY'],
  llamafile: ['LLAMAFILE_*'],
  manus: ['MANUS_*'],
  'mistral-vibe': ['MISTRAL_API_KEY'],
  ollama: ['OLLAMA_*'],
  'open-interpreter': ['INTERPRETER_*', 'OPENAI_API_KEY'],
  'qwen-code': ['QWEN_API_KEY', 'DASHSCOPE_API_KEY'],
  rowboatx: ['ROWBOATX_*'],
  rovo: ['ROVO_*'],
  'shell-pilot': ['SHELL_PILOT_*'],
  smithery: ['SMITHERY_*'],
  trae: ['TRAE_*'],
};

class EnvironmentManager {
  private sessions: Map<string, SessionEnvironment> = new Map();
  private globalOverrides: Record<string, string> = {};
  private globalSecrets: Set<string> = new Set();

  constructor() {}

  setGlobalOverride(key: string, value: string): void {
    this.globalOverrides[key] = value;
  }

  removeGlobalOverride(key: string): void {
    delete this.globalOverrides[key];
  }

  addGlobalSecret(key: string): void {
    this.globalSecrets.add(key);
  }

  createSessionEnvironment(
    sessionId: string,
    cliType: CLIType,
    config?: Partial<EnvironmentConfig>
  ): Record<string, string> {
    const envConfig: EnvironmentConfig = {
      inherit: config?.inherit ?? true,
      variables: config?.variables ?? {},
      secrets: config?.secrets ?? [],
      passthrough: config?.passthrough ?? [],
    };

    const resolved = this.resolveEnvironment(cliType, envConfig);

    this.sessions.set(sessionId, {
      sessionId,
      cliType,
      config: envConfig,
      resolved,
    });

    return resolved;
  }

  private resolveEnvironment(cliType: CLIType, config: EnvironmentConfig): Record<string, string> {
    const result: Record<string, string> = {};

    if (config.inherit) {
      // Safely handle missing CLI types if any future ones are added without updating map
      const cliSpecific = CLI_SPECIFIC_VARS[cliType] || [];
      const passthroughPatterns = [
        ...DEFAULT_PASSTHROUGH_VARS,
        ...config.passthrough,
        ...cliSpecific,
      ];

      for (const [key, value] of Object.entries(process.env)) {
        if (value && this.matchesPattern(key, passthroughPatterns)) {
          result[key] = value;
        }
      }
    }

    for (const [key, value] of Object.entries(this.globalOverrides)) {
      result[key] = value;
    }

    for (const [key, value] of Object.entries(config.variables)) {
      result[key] = this.expandVariables(value, result);
    }

    return result;
  }

  private matchesPattern(key: string, patterns: string[]): boolean {
    for (const pattern of patterns) {
      if (pattern.endsWith('*')) {
        const prefix = pattern.slice(0, -1);
        if (key.startsWith(prefix)) return true;
      } else if (pattern === key) {
        return true;
      }
    }
    return false;
  }

  private expandVariables(value: string, env: Record<string, string>): string {
    return value.replace(/\$\{(\w+)\}/g, (_, varName) => {
      return env[varName] || process.env[varName] || '';
    });
  }

  getSessionEnvironment(sessionId: string): Record<string, string> | undefined {
    return this.sessions.get(sessionId)?.resolved;
  }

  updateSessionVariable(sessionId: string, key: string, value: string): void {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    session.config.variables[key] = value;
    session.resolved[key] = this.expandVariables(value, session.resolved);
  }

  removeSessionVariable(sessionId: string, key: string): void {
    const session = this.sessions.get(sessionId);
    if (!session) return;

    delete session.config.variables[key];
    delete session.resolved[key];
  }

  deleteSessionEnvironment(sessionId: string): void {
    this.sessions.delete(sessionId);
  }

  getSanitizedEnvironment(sessionId: string): Record<string, string> {
    const session = this.sessions.get(sessionId);
    if (!session) return {};

    const result: Record<string, string> = {};
    const secretKeys = new Set([...session.config.secrets, ...this.globalSecrets]);

    for (const [key, value] of Object.entries(session.resolved)) {
      if (secretKeys.has(key) || this.looksLikeSecret(key)) {
        result[key] = '***REDACTED***';
      } else {
        result[key] = value;
      }
    }

    return result;
  }

  private looksLikeSecret(key: string): boolean {
    const secretPatterns = [
      /api[_-]?key/i,
      /secret/i,
      /password/i,
      /token/i,
      /credential/i,
      /auth/i,
      /private[_-]?key/i,
    ];
    return secretPatterns.some(pattern => pattern.test(key));
  }

  getRequiredVarsForCLI(cliType: CLIType): string[] {
    const required: Record<CLIType, string[]> = {
      opencode: [],
      claude: ['ANTHROPIC_API_KEY'],
      aider: ['OPENAI_API_KEY'],
      cursor: [],
      continue: [],
      cody: ['SRC_ACCESS_TOKEN'],
      copilot: ['GITHUB_TOKEN'],
      custom: [],
      adrenaline: [],
      'amazon-q': [],
      'amazon-q-developer': [],
      'amp-code': [],
      auggie: [],
      'azure-openai': ['AZURE_OPENAI_API_KEY', 'AZURE_OPENAI_ENDPOINT'],
      bito: [],
      byterover: [],
      'claude-code': ['ANTHROPIC_API_KEY'],
      'code-codex': [],
      codebuff: [],
      codemachine: [],
      codex: [],
      crush: [],
      dolt: [],
      factory: [],
      gemini: ['GEMINI_API_KEY'],
      goose: [],
      grok: ['GROK_API_KEY'],
      jules: [],
      'kilo-code': [],
      kimi: ['KIMI_API_KEY'],
      llm: [],
      litellm: [],
      llamafile: [],
      manus: [],
      'mistral-vibe': ['MISTRAL_API_KEY'],
      ollama: [],
      'open-interpreter': [],
      'qwen-code': ['QWEN_API_KEY'],
      rowboatx: [],
      rovo: [],
      'shell-pilot': [],
      smithery: [],
      trae: [],
    };

    return required[cliType] || [];
  }

  validateEnvironmentForCLI(cliType: CLIType, env: Record<string, string>): { valid: boolean; missing: string[] } {
    const required = this.getRequiredVarsForCLI(cliType);
    const missing = required.filter(key => !env[key]);
    return { valid: missing.length === 0, missing };
  }
}

export const environmentManager = new EnvironmentManager();

File: go-port/pkg/server/services/env/environment_manager.go
package env

import (
	"os"
	"strings"
	"sync"


	"borg-orchestrator/pkg/shared"
)

type EnvironmentConfig struct {
	Inherit     bool              `json:"inherit"`
	Variables   map[string]string `json:"variables"`
	Secrets     []string          `json:"secrets"`
	Passthrough []string          `json:"passthrough"`
}

type SessionEnvironment struct {
	SessionID string            `json:"sessionId"`
	CLIType   shared.CLIType    `json:"cliType"`
	Config    EnvironmentConfig `json:"config"`
	Resolved  map[string]string `json:"resolved"`
}

var DefaultPassthroughVars = []string{
	"PATH", "HOME", "USER", "SHELL", "TERM",
	"LANG", "LC_ALL", "TMPDIR", "TMP", "TEMP",
	"NODE_ENV", "BUN_ENV",
}

var CommonSecrets = []string{
	"OPENAI_API_KEY",
	"ANTHROPIC_API_KEY",
	"GEMINI_API_KEY",
	"DEEPSEEK_API_KEY",
	"XAI_API_KEY",
	"GITHUB_TOKEN",
	"GITLAB_TOKEN",
	"NPM_TOKEN",
	"AWS_ACCESS_KEY_ID",
	"AWS_SECRET_ACCESS_KEY",
	"STRIPE_SECRET_KEY",
}

type EnvironmentManagerService struct {
	sessions map[string]*SessionEnvironment
	global   EnvironmentConfig
	mu       sync.RWMutex
}

func NewEnvironmentManagerService() *EnvironmentManagerService {
	return &EnvironmentManagerService{
		sessions: make(map[string]*SessionEnvironment),
		global: EnvironmentConfig{
			Inherit:     true,
			Variables:   make(map[string]string),
			Secrets:     append([]string{}, CommonSecrets...),
			Passthrough: append([]string{}, DefaultPassthroughVars...),
		},
	}
}

func (s *EnvironmentManagerService) ConfigureGlobal(config EnvironmentConfig) {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.global.Inherit = config.Inherit

	if config.Variables != nil {
		s.global.Variables = config.Variables
	}

	if config.Secrets != nil {
		s.global.Secrets = append([]string{}, config.Secrets...)
	}

	if config.Passthrough != nil {
		s.global.Passthrough = append([]string{}, config.Passthrough...)
	}
}

func (s *EnvironmentManagerService) PrepareEnvironment(sessionID string, cliType shared.CLIType, customConfig *EnvironmentConfig) (*SessionEnvironment, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	config := EnvironmentConfig{
		Inherit:     s.global.Inherit,
		Variables:   make(map[string]string),
		Secrets:     append([]string{}, s.global.Secrets...),
		Passthrough: append([]string{}, s.global.Passthrough...),
	}

	for k, v := range s.global.Variables {
		config.Variables[k] = v
	}

	if customConfig != nil {
		config.Inherit = customConfig.Inherit
		for k, v := range customConfig.Variables {
			config.Variables[k] = v
		}
		if customConfig.Secrets != nil {
			config.Secrets = append(config.Secrets, customConfig.Secrets...)
		}
		if customConfig.Passthrough != nil {
			config.Passthrough = append(config.Passthrough, customConfig.Passthrough...)
		}
	}

	resolved := make(map[string]string)

	if config.Inherit {
		for _, env := range os.Environ() {
			parts := strings.SplitN(env, "=", 2)
			if len(parts) == 2 {
				resolved[parts[0]] = parts[1]
			}
		}
	} else {
		for _, key := range config.Passthrough {
			if val, ok := os.LookupEnv(key); ok {
				resolved[key] = val
			}
		}
	}

	for k, v := range config.Variables {
		resolved[k] = v
	}

	env := &SessionEnvironment{
		SessionID: sessionID,
		CLIType:   cliType,
		Config:    config,
		Resolved:  resolved,
	}

	s.sessions[sessionID] = env
	return env, nil
}

func (s *EnvironmentManagerService) GetEnvironment(sessionID string) (*SessionEnvironment, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	env, ok := s.sessions[sessionID]
	return env, ok
}

func (s *EnvironmentManagerService) UpdateSessionVariable(sessionID string, key, value string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	env, ok := s.sessions[sessionID]
	if !ok {
		return false
	}

	env.Config.Variables[key] = value
	env.Resolved[key] = value
	return true
}

func (s *EnvironmentManagerService) RemoveSession(sessionID string) {
	s.mu.Lock()
	defer s.mu.Unlock()

	delete(s.sessions, sessionID)
}

func (s *EnvironmentManagerService) IsSecret(key string, sessionID string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	checkSecret := func(secrets []string, k string) bool {
		for _, secret := range secrets {
			if secret == k || (strings.HasSuffix(secret, "*") && strings.HasPrefix(k, strings.TrimSuffix(secret, "*"))) {
				return true
			}
		}
		return false
	}

	if checkSecret(s.global.Secrets, key) {
		return true
	}

	if sessionID != "" {
		if env, ok := s.sessions[sessionID]; ok {
			if checkSecret(env.Config.Secrets, key) {
				return true
			}
		}
	}

	return false
}

func (s *EnvironmentManagerService) RedactLog(log string, sessionID string) string {
	s.mu.RLock()
	defer s.mu.RUnlock()

	secretsToRedact := []string{}

	addSecrets := func(resolved map[string]string, secrets []string) {
		for k, v := range resolved {
			if v != "" && len(v) > 3 {
				for _, secret := range secrets {
					if secret == k || (strings.HasSuffix(secret, "*") && strings.HasPrefix(k, strings.TrimSuffix(secret, "*"))) {
						secretsToRedact = append(secretsToRedact, v)
					}
				}
			}
		}
	}

	// add globals
	globalResolved := make(map[string]string)
	for _, env := range os.Environ() {
		parts := strings.SplitN(env, "=", 2)
		if len(parts) == 2 {
			globalResolved[parts[0]] = parts[1]
		}
	}
	for k, v := range s.global.Variables {
		globalResolved[k] = v
	}
	addSecrets(globalResolved, s.global.Secrets)

	if sessionID != "" {
		if env, ok := s.sessions[sessionID]; ok {
			addSecrets(env.Resolved, env.Config.Secrets)
		}
	}

	redacted := log
	for _, secret := range secretsToRedact {
		redacted = strings.ReplaceAll(redacted, secret, "[REDACTED]")
	}

	return redacted
}
